### ISSUE TITLE
unexpected close connection message

### ISSUE THREAD

**Issue by Pu4** [aspnet/SignalR#2207]
My project is .Net Core 2.0 Web API. It runs as stateless service in service fabric both locally and in Azure. For some time I've been using signalr v 1.0.0-alpha2-final both locally and on demo in azure and it worked. But yesterday I had to update it to latest 1.0.0-preview2-final. It works locally but after deployment on azure server closes connection immediately after it's being established. For testing I'm using js client. Package @aspnet/signalr@1.0.0-preview2-final. I've tried different session persistent setting of load balancer, tried to use c# client with same signalr package version that api has but nothing helped.

Server-side code

```
public void ConfigureServices(IServiceCollection services)
{
......
var corsPolicy = new Microsoft.AspNetCore.Cors.Infrastructure.CorsPolicy();
corsPolicy.Headers.Add("*");
corsPolicy.Methods.Add("*");
corsPolicy.Origins.Add("*");
corsPolicy.SupportsCredentials = true;

services.AddCors(x => x.AddPolicy("corsGlobalPolicy", corsPolicy));

services.AddSignalR();
......
}


public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
.......
app.UseCors("corsGlobalPolicy");
app.UseSignalR(routes =>
{
routes.MapHub<UserNotification>("/Notification");
});
......
}
```

Client side

```
let connection = new signalR.HubConnection('http://{domain}/Notification');
....
//Some event handlers here
....
connection.start();
```

Console logs
![dbjdl](https://user-images.githubusercontent.com/11652519/39664277-8095c0ec-5089-11e8-8e59-e10ae65b1822.png)

**Comment by davidfowl** [comment]
Try turning on detailed errors on the server.

https://github.com/aspnet/SignalR/blob/32b4d5cc6ce8a938d83d473a7e0030184b008e4a/test/Microsoft.AspNetCore.SignalR.Tests/Startup.cs#L17-L20

**Comment by davidfowl** [comment]
Ah good! More details. What does your Hub look like?

**Comment by Pu4** [comment]
```
[Authorize(AuthenticationSchemes = "Bearer")]
    public class UserNotification : Hub
    {
        private readonly ITransactionsService _transactionService;
        private readonly IConsumersService _consumersService;
        private readonly IConfiguration _config;
        private readonly IDataEncryptor _dataEncryptor;
        private readonly IRsaEncryptor _rsaEncryptor;
        private IHubContext<UserNotification> _context;
        private MessageReceiver _receiver;
        private int invocationCount = 0;
        private ServiceBusTransactionListener _listener;
        private static Dictionary<string, EncryptorKey> clientEncryptorKeys = new Dictionary<string, EncryptorKey>();
        public UserNotification(ITransactionsService transactionService,
            IConsumersService consumersService,
            IConfiguration config,
            IHubContext<UserNotification> context,
            IDataEncryptor dataEncryptor,
            ServiceBusTransactionListener listener,
            IRsaEncryptor rsaEncryptor)
        {
            _transactionService = transactionService;
            _consumersService = consumersService;
            _config = config;
            _context = context;
            _receiver = new MessageReceiver(_config.GetConnectionString("ServiceBus"),
                _config["ServiceBus:TopicName"] + "/Subscriptions/APISubscription");
            _listener = listener;
            _dataEncryptor = dataEncryptor;
            _rsaEncryptor = rsaEncryptor;
        }
        private Task LogMessageHandlerException(ExceptionReceivedEventArgs e)
        {
            Console.WriteLine("Exception: \"{0}\" {0}", e.Exception.Message, e.ExceptionReceivedContext.EntityPath);
            return Task.CompletedTask;
        }

        
        public async Task Send()
        {
            Transaction t = new Transaction()
            {
                ToClientId = "mobile_client"
            };
            Message m = new Message
            {
                ContentType = "application/json",
                Body = Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(t)),
                To = t.ClientId
            };
            string ep = _config["ServiceBus:TopicName"];
            var sender = new MessageSender(_config.GetConnectionString("ServiceBus"), ep);
            await sender.SendAsync(m);
        }

        public async Task GetMessages()
        {
            var m = await _receiver.PeekAsync(10);
            await Clients.Caller.SendAsync(nameof(SignalRMethods.GetMessages), m);

        }

        /// <summary>
        /// Send server public key and nonce 
        /// create symmitric key by client public key.
        /// <returns></returns>
        public async Task PublicKeyExchange(string clientPublicKeyBytesString)
        {
            // var clientPublicKeyBytes = clientPublicKeyBytesString.Split(',').Select(x => Byte.Parse(x)).ToArray();
            //ToDo remove test value of clientPublicKeyBytes          
            //var testEncryptorKey = _dataEncryptor.CreateX25519EncryptorKey();
            //clientPublicKeyBytes = _dataEncryptor.GetX25519EncryptorPublicKeyBytes(testEncryptorKey.Key);

            var clientPublicKeyBytes = System.Convert.FromBase64String(clientPublicKeyBytesString);

            // Decrypt client public key with ssl sertificate
            clientPublicKeyBytes = _rsaEncryptor.DecryptRsa(clientPublicKeyBytes);

            var clientId = GetCurrentClientId();
            var encryptorKey = _dataEncryptor.CreateX25519EncryptorKey();

            var publicKey = _dataEncryptor.GetX25519EncryptorPublicKeyBytes(encryptorKey.Key);

            //Encrypt public key with SSL sertificate
            publicKey = _rsaEncryptor.EncryptRsa(publicKey);

            var model = new ServerPublicKeyDto
            {
                PublicKey = System.Convert.ToBase64String(publicKey),
                Nonce = System.Convert.ToBase64String(_dataEncryptor.GetNonceBytes(encryptorKey.Nonce))
            };

            encryptorKey.Key = _dataEncryptor.CreateAes256GcmSymmetricKey(clientPublicKeyBytes, encryptorKey.Key);
            if (clientEncryptorKeys.ContainsKey(clientId))
                clientEncryptorKeys.Remove(clientId);
            clientEncryptorKeys.Add(clientId, encryptorKey);
            await Clients.Caller.SendAsync(nameof(SignalRMethods.PublicKeyExchange), JsonConvert.SerializeObject(model));
        }

        /// <summary>
        /// Mark all of the current user's requests as 'viewed'. 
        /// Note this method is only available for CONSUMER.
        /// <returns></returns>
        public async Task ConsumerRequestsHistoryViewPut()
        {
            string data = string.Empty;
            var result = false;

            try
            {
                await _transactionService.MarkTransactionsAsViewedAsync(GetCurrentClientId());
                result = true;
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.MarkTransactionsAsViewed, GetCurrentClientId(), Context, null, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsHistoryViewPut), data);
        }

        /// <summary>
        /// Get number of not viewed requests in cosumers history.
        /// Count all requests that are not marked as 'viewed' in consumer's request history. Note this method is only available for CONSUMER.
        /// </summary>
        public async Task ConsumerRequestsHistoryViewGet()
        {
            string data = String.Empty;
            var result = new CountResponce();

            try
            {
                result = new CountResponce
                {
                    Count = await _transactionService.GetCountViewedTransactionAsync(GetCurrentClientId())
                };

                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetCountViewedRequests, GetCurrentClientId(), Context, null, result);

            var encryptorKey = clientEncryptorKeys.GetValueOrDefault(GetCurrentClientId());
            //ToDo return error if encryptorKey == null
            var chiperText = _dataEncryptor.EncryptDataByAes256Gcm(data, encryptorKey);
            //ToDo remove decrypt method, from dataEncryptor as well. Only for testing
            var newText = _dataEncryptor.Decrypt(chiperText, encryptorKey);
            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsHistoryViewGet), System.Convert.ToBase64String(chiperText));
        }

        /// <summary>
        /// Get information about current consumer.
        /// Not this method is only available for CONSUMER.
        /// </summary>
        public async Task ConsumerGet()
        {
            string clientId = GetCurrentClientId();
            var data = String.Empty;
            var result = new ConsumerDto();

            try
            {
                result = await _consumersService.GetCurrentUser(clientId);
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetInformationAboutYourself, clientId, Context, null, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerGet), data);
        }

        /// <summary>
        /// Get list of merchants of current consumer.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        public async Task ConsumerMerchantsGet()
        {
            string clientId = GetCurrentClientId();
            var data = string.Empty;
            var result = new ConsumerMerchantsDto();

            try
            {
                result = await _consumersService.GetUserMerchantsAsync(clientId);
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetMerchantsForUser, clientId, Context, null, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerMerchantsGet), data);
        }

        /// <summary>
        /// Get list of validation request of current user for given merchant.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="companyId">Company identifier</param>
        /// <returns></returns>
        public async Task ConsumerMerchantsSeedRequestsGet(string companyId)
        {

            var clientId = GetCurrentClientId();
            var data = string.Empty;
            var result = new List<RequestViewModel>();

            try
            {
                var decryptFromBade64Data = System.Convert.FromBase64String(companyId);
                companyId = _dataEncryptor.Decrypt(decryptFromBade64Data, clientEncryptorKeys.GetValueOrDefault(GetCurrentClientId()));

                result = await _consumersService.GetAllMerchantTransactionsForUser(companyId, clientId);
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetMerchantTransactionsForUser, clientId, Context, companyId, result);

            var encryptorKey = clientEncryptorKeys.GetValueOrDefault(GetCurrentClientId());
            //ToDo return error if encryptorKey == null
            var chiperText = _dataEncryptor.EncryptDataByAes256Gcm(data, encryptorKey);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerMerchantsSeedRequestsGet), System.Convert.ToBase64String(chiperText));
        }

        /// <summary>
        /// Update information about current consumer.
        /// Not this method is only available for CONSUMER.
        /// </summary>
        /// <param name="user">Information about user</param>
        /// <returns></returns>
        public async Task ConsumerPut(ConsumerDto model)
        {
            var data = string.Empty;
            var result = new ConsumerDto();

            try
            {
                result = await _consumersService.UpdateConsumerInfo(model);
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetMerchantsForUser, GetCurrentClientId(), Context, model, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerPut), data);
        }

        /// <summary>
        /// Get list of validation request of current user. 
        /// If status is specified then requests will be found in accordance to status otherwise all requests will be found.
        /// This method support pagging. So you may use pageNumber and pageSize parameters.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="status">Status of transaction (Active, Expired, Accepted, Rejected)</param>
        /// <param name="pageNumber">Page number</param>
        /// <param name="pageSize">Page size</param>
        public async Task ConsumerRequestsGet(string status, int pageNumber = 1, int pageSize = 10)
        {
            var clientId = GetCurrentClientId();
            var data = string.Empty;
            var result = new List<RequestViewModel>();

            try
            {
                result = await _transactionService.GetAllTransactionAsync(status, pageNumber, pageSize, clientId);
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            var inputData = new { Status = status, PageNumber = pageNumber, PageSize = pageSize };
            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetTransactions, clientId, Context, inputData, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsGet), data);
        }

        /// <summary>
        /// Get list of history requests of current user. It will be expired requests or status is not ACTIVE.
        /// This method support pagging. So you may use pageNumber and pageSize parameters.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="pageNumber"></param>
        /// <param name="pageSize"></param>
        public async Task ConsumerRequestsHistoryGet(int pageNumber = 1, int pageSize = 10)
        {
            var clientId = GetCurrentClientId();
            var data = string.Empty;
            var result = new List<RequestViewModel>();

            try
            {
                result = await _transactionService.GetHistoryTransactions(pageNumber, pageSize, clientId);
                data = JsonConvert.SerializeObject(result);
            }
            catch (Exception ex)
            {

                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetHistory, clientId, Context, new { PageNumber = pageNumber, PageSize = pageSize }, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsHistoryGet), data);
        }

        /// <summary>
        /// Accept validation request with give id.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="id">Transaction identifier</param>
        /// <returns></returns>
        public async Task ConsumerRequestsIdAcceptGet(string id)
        {
            var data = string.Empty;
            RequestViewModel result = null;

            try
            {
                result = await _transactionService.UpdateTransactionStatusAsync(id, CosmosDbLibrary.Enums.Status.Accepted);
                data = result != null ? JsonConvert.SerializeObject(result) : JsonConvert.SerializeObject(new { message = "Transaction not found" });
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.AcceptRequest, GetCurrentClientId(), Context, id, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsIdAcceptGet), data);


        }

        /// <summary>
        /// Reject validation request with given id.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="id">Transaction identifier</param>
        public async Task ConsumerRequestsIdAcceptDelete(string id)
        {
            var data = string.Empty;
            RequestViewModel result = null;

            try
            {
                result = await _transactionService.UpdateTransactionStatusAsync(id, CosmosDbLibrary.Enums.Status.Rejected);
                data = result != null ? JsonConvert.SerializeObject(result) : JsonConvert.SerializeObject(new { message = "Transaction not found" });
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.RejectRequest, GetCurrentClientId(), Context, id, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsIdAcceptDelete), data);

        }
        //ToDo include clientId and companyId as filter parameters

        /// <summary>
        /// Get information about validation request with given id.
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="transactionId">Transaction identifier</param>
        /// <param name="companyId">Company identifier</param>
        /// <returns></returns>
        public async Task ConsumerRequestsIdGet(string transactionId, string companyId)
        {
            var clientId = GetCurrentClientId();
            var data = string.Empty;
            var result = new RequestViewModel();

            try
            {
                result = await _transactionService.GetCompanyTransactionByIdAsync(transactionId, companyId);
                data = result != null ? JsonConvert.SerializeObject(result) : JsonConvert.SerializeObject(new { message = "Transaction not found" });
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.GetTransactionById, clientId, Context, new { transactionId, companyId }, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsIdGet), data);
        }

        /// <summary>
        /// Delete specified list of requests from history of current user. 
        /// Note this method is only available for CONSUMER.
        /// </summary>
        /// <param name="model">Object with transactions identifiers array</param>
        /// <returns></returns>
        public async Task ConsumerRequestsHistoryPut(TransactionsDeleteInputModel model)
        {
            var data = string.Empty;
            TransactionsDeleteViewModel result = null;

            try
            {
                result = await _transactionService.DeleteHistoryTransactions(model, GetCurrentClientId());
                if (result != null && result.ErrorCollector.Count > 0)
                {
                    data = JsonConvert.SerializeObject(new { errors = result.ErrorCollector });

                }
                else if (result == null)
                {
                    data = JsonConvert.SerializeObject(new { message = "Transactions not found" });
                }
                else
                {
                    data = JsonConvert.SerializeObject(new { result = "ok" });
                }
            }
            catch (Exception ex)
            {
                await SignalRExceptionsLogHelper.LogException(ex, _config);
                data = JsonConvert.SerializeObject(new { error = ex.Message });
            }

            await SignalRAuditHelper.AuditEvent(AuditPrivateWebEventsConstants.AcceptRequest, GetCurrentClientId(), Context, model, result);

            await Clients.Caller.SendAsync(nameof(SignalRMethods.ConsumerRequestsHistoryPut), data);
        }

        
        private string GetCurrentClientId()
        {
            return Context.User.FindFirst("client_id").Value;

        }
```

**Comment by davidfowl** [comment]
Unfortunately, we don't have a stack trace if you attach the debugger to the server and catch all exceptions, it'll probably show you where the problem is. 

It looks like it's failing before you invoke any methods so maybe this is the problem:

```C#
_receiver = new MessageReceiver(_config.GetConnectionString("ServiceBus"),
                _config["ServiceBus:TopicName"] + "/Subscriptions/APISubscription");
```

That's just a guess. 

Also some comments about the code, Hubs are transient and created *PER Invocation*, what that means is you'll end up creating a new `MessageReceiver` per invocation which I'm betting you don't want to do.

**Comment by Pu4** [comment]
Thank you for your advice. I'm aware about hub's life cycle. Method which uses `MessageReceiver` is not for production and wasn't included in listing. It will be removed shortly from hub with `MessageReceiver` itself.
As for my initial question, problem was solved by updating signlR to 1.0.0-rc1-final.

**Comment by davidfowl** [comment]
Cool
